### ISSUE TITLE
meanings of digits in semantic version

### ISSUE THREAD

**Issue by keighrim** [clamsproject/mmif#14]
As a subtask of #12, I'd like to specify what three digits mean in the version space of MMIF. 

### Versioning principles 
It is preferred that versions of sub-components are tied to each other. In an ideal world, all versions are exactly synchronized, but as we discuss here in the below, that is not very practically likely. 
### Type of changes and using digits 
* version number format `x.y.z`: version numbers consist of three digits, each digit is positive integer
* **major**: major changes are breaking changes that are not backward compatible - any major change will increase `x` digit in the version number *by 1*. 
* **minor**: minor changes add features - any minor change will increase `y` digit
* **patch**: patches are not adding any feature - any patch will increase `z` digit

### MMIF components 
As described in #12, there are largely 5 sub-components; 
1. MMIF specification (`spec` hereinafter)
    * either as a set of abstract API, or simply as documentation
1. MMIF JSON schema (`schema` hereinafter)
1. MMIF JSON-LD context (`context` hereinafter)
1. MMIF reserved vocabulary  (`vocab` hereinafter)
1. MMIF serialization API implementation (`sdk` hereinafter) 

### td; dr
* updates in `schema` are most likely major
* updates in `vocab` are most likely minor (additions) 
* updates in `sdk` are always patches, but not reflected to main `spec` version
* no standalone updates in `context`
* all five sub-components share `x` and `y` digtis
* `spec`, `schema`, `context`, and `vocab` share `z` as well
* `sdk` never make major or minor changes

#### Changes in `spec`
Any changes in a sub-component would result in change in specification as well, and thus will increase the version number. Changes in specification that are not accompanied by changes in component(s) are not very imaginable, unless changes to fix typographical errors. Fixing typos should be a *patch* level version bump. 

#### Changes in `schema` 
Possible changes include (not limited to) 
* addition of fields (e.g. introducing new metadata) 
* re-organization of existing fields (e.g. changes in value types)
* deletion of fields (deprecation) 

In terms of automatic JSON validation, an addition of a field/term can be non-breaking change unless the new field is *required*. All other changes are not backward compatible.

#### Changes in `vocab` 
As there is not a *forcible* way of validating semantics of MMIF string using `vocab` as validation source, *backward compatibility* must mean something different here. However, I'd like to suggest all additions (items, props of an item) to the ontology must be non-breaking and *minor* level version bump-ups. 

#### Changes in `context` 
Any addition of items in the `vocab` or `schema` should update `context` as well. But changes originating from `context` is not so likely. 

#### Changes in `sdk`
In contrast to other components, `sdk` is actual code, and is very prone to have software bugs. However, as `sdk` is a mere implementation of what's defined and documented in the `spec`, unless it's originated from updates of other components, all standalone changes in `sdk` would be only fixing bugs, optimizing code, or adding supplementary (possibly undocumented) APIs. So those types of updates won't add (or change) any *feature* to the MMIF `spec`. But, by its virtue, `sdk` will receive a lot more *patches* compared to other components, thus I propose we untie patch number (`z`) of `sdk` from MMIF version. 

### Version binding 
In conclusion, we want to stick to the principle of shared version numbers between all relevant components, however, because `sdk` is more likely to receive frequent patches, patch number of `sdk` needs to be independent from MMIF version. 

### Handling MMIF versions in consumers
We don't want to pose any restriction on consumer (e.g. [MMIF visualizer](https://github.com/clamsproject/mmif-visualizer)) implementation. By publishing `sdk` to version-controlled package repositories (e.g. pypi), we can expect a specific implementation of a consumer is tied to a specific version of MMIF (at least on `x` and `y` level). If a developer of such a consumer wants to support multiple versions of MMIF using our `sdk`, it might not be trivial as the `sdk` does not inherently support backward compatibility (by design), so the recommended way of support many versions would be maintaining different versions of the consumer software - which is not ideal (in fact a really bad practice, I believe). This leads us to a new question that might need its own issue card; namely, if it turned out that `sdk` of older version (say 1.2.18) contains some critical bug after the main MMIF version went up with some major changes (say latest MMIF version is 2.1.0), do we need to fix the old `sdk` and publish a new version (say 1.2.19)?

**Comment by keighrim** [comment]
To answer the above question; yes we probably need to patch old version of the SDK if some critical bugs are found. For now, and for python particularly, we can exploit [the `post-release` versioning](https://www.python.org/dev/peps/pep-0440/#post-releases) as officially proposed.

**Comment by keighrim** [comment]
Although this thread started as a subtask of #12 (which is now closed), it contains basic guidelines for MMIF developers, and hence let's keep it open at least until we implement all *builders*.

**Comment by marcverhagen** [comment]
Need to rethink this now that we have split off the python-mmif code.

**Comment by keighrim** [comment]
Here's my attempt to update the discussion above, quoting @marcverhagen 's remarks from clamsproject/mmif-python#163 

> 1. The JSON schema are different. This breaks the syntax and could, and probably even should, throw an error. We agreed that this could be the thing that gets us to bump up the major version once we finally hit version 1.0.0. A complication here is that some changes will invalidate things that used to be okay and others won't. For example, making a property optional won't break anything, making it required would, at least for all the cases where we do not have that property in the MMIF file.

### So regarding changes in `schema`
We have breaking and non-breaking changes. Namely, 
- *breaking*: any changes in *required* fields (addition, removal, internal modification) 
  - get +1 `major` bump once 1.0.0 is published, a `minor` bump by then.
- *non-breaking*: any changes in optional fields (addition, removal, internal modification) 
  - get +1 `minor` bump once 1.0.0 is published, a `patch` bump by then. 

> 2. We add, remove or change an annotation type (where changing means adding, removing or changing a property or metadata property). We had been thinking of these as changes to the minor version. Here we could glue together a chain with and EAST version that uses MMIF 1.2.0 and Tesseract that uses MMIF 1.3.0. This probably won't break the chain unless an app is set up to throw an error when required input isn't there (and I think the app should in that case set up to be really specific about what kind of input it misses and put that in the error message). Glueing together these components may break the intended result if there are enough changes in MMIF pertinent to the annotation types that EAST and Tesseract use. For example, Tesseract would return an empty result because it did not find and bounding boxes of the right version.
 
### So regarding changes in `vocab`
Again, we have breaking and non-breaking changes. 
- *breaking*: removing, renaming, or changing nested structures of metadata or property of existing types
  - note that there are only a limited number of properties that have nested structures (e.g. `coordinates` of `Polygon` type), most of properties are simple strings
  - these kinds of changes might occur much rarely once we hit the first stable version (1.0.0)
  - get +1 `major` bump once 1.0.0 is published, a `minor` bump by then.
- *non-breaking*: addition of properties to existing types, or addition of new types
  - get +1 `patch` bump once 1.0.0 is published, a `patch` bump by then. 
Before we move on to next component, I want to remind us all that (as mentioned above), with current implementation of `vocabulary` package in the `SDK` (which is simply a set of python enums) there's no way to actually *validate* annotation properties in the input MMIF file. So even with *breaking* changes, the code might just run (and probably fails). 

### And then regarding changes in `spec` 
The situation is not that different here. Any changes in a `schema`/`vocab` would result in change in specification as well, and thus will increase the version number accordingly. Changes in specification that are not accompanied by changes in other component(s) are not very imaginable, unless changes to fix typographical errors or clausal clarification. Errata and clausal clarification should be `patch` level version bumps. 

### Then on version binding
I think we still agree, after the separation of the SDK codebase, that we use a single version number for all components in specification (spec, vocab, schema) and `major` and `minor` of the SDK version are bound to the spec version. With that being said, 
> - App metadata never refer to the version of a required annotation type, which is as it should be, the question is whether the MMIF version used by the app not only determines what the app creates but also what it needs as input.
> - I don't like at all the idea of all applications needing to be changed following a minor change in the specifications, even if those changes to not impact the app at all. This would be unmanageable.

The app metadata proposed in clamsproject/clams-python#53 adds `mmif_version` to the metadata. So, we can use it to determine the compatibility of an input MMIF, accepting input that matches `major` level version. Thus this should solve the second problem above. To that end 
> - The `clams-python` package is set up to require a specific X.Y.Z version of the `mmif-python` package. I don't know what exactly the nature of this dependency is. Would it be feasible to use a range? At least for the patch-level and maybe also for the minor version.

When the mmif-python and clams-python SDK are updated so that a CLAMS app can accept an input MMIF that's different in `minor` or `patch` level, the python dependency should not be relevant at all. And from the engineering perspective, I don't like the idea of having a range of versions as dependency. But we can re-consider this at any point. 

### Finally on forward compatibility
Just to clarify the nature of version control between `mmif-python` and `clams-python`, `mmif-python` simply keeps the version of the spec as `mmif.__specver__` and `clams-python` re-imports that as `clams.__specver__`. So with the change proposed at clamsproject/mmif-python#163, if a CLAMS that is based on `clams-python==2.2.3` which in turn relies on `mmif.__version__==3.2.10 && mmif.__specver__==3.2.3` should be able to process any `3.[0-2].b` MMIF inputs. 

At this point, we have to think about forward compatibility. Namely, must this app be able to process a, for example, `3.7.x` MMIF input? Note that with proposal in this comment, once we reach 1.0.0, the only place the `minor` number gets a bump is *non-breaking* changes in `schema`. So the `schema==3.7.x` might have any number of optional fields added. But the `mmif-python==3.2.x` probably don't know how to handle those added fields, and possibly break with the higher version input. The issue gets a bit more complicated when it comes to `0.x.y`, as within this range, `minor` and `patch` levels are conceptually mixed.

#### p.s.
> - Maybe `ClamsApp.annotate()` in `clams-python` could throw an error when the app's major MMIF version conflicts with the major MMIF version in the input MMIF file.

This should be a separate issue in the clams-python repository.

**Comment by keighrim** [comment]
If no one opposes the above proposal, I'll start implementing https://github.com/clamsproject/mmif-python/issues/163 based on the proposal.

**Comment by marcverhagen** [comment]
Ah, I knew I was forgetting something. I had some comments on the proposal and will add them in asap.

**Comment by keighrim** [comment]
While working on https://github.com/clamsproject/mmif-python/pull/170, I realized that the following is not true. 
>  all standalone changes in sdk would be only fixing bugs, optimizing code, or adding supplementary (possibly undocumented) APIs ... sdk will receive a lot more patches 

Namely standalone changes in SDK can also be a breaking one (not breaking the spec but only SDK API's). So with that in mind, if we continue to *tie-up* major and minor versions between spec and SDK (as in the above example), that could mean that self-breaking changes in the SDK will still get a patch bump.

**Comment by marcverhagen** [comment]
> If no one opposes the above proposal, I'll start implementing [clamsproject/mmif-python#163](https://github.com/clamsproject/mmif-python/issues/163) based on the proposal.

Just out of curiosity, what would you implement? Isn't this about how we choose what kind of version update to use given things we put in it?

**Comment by keighrim** [comment]
Yeah, https://github.com/clamsproject/mmif-python/issues/163 proposed disregarding the patch level when comparing two `@type` URIs, but the above proposes to disregard patch and minor when doing so. 

I think most of the comparison logic is already implemented in https://github.com/clamsproject/mmif-python/pull/170. I'll write more comprehensive test cases to see if that PR actually working as intended with a pipeline with version-mixed MMIF ins and outs.

**Comment by marcverhagen** [comment]
### Version binding

> The app metadata proposed in clamsproject/clams-python#53 adds `mmif_version` to the metadata. So, we can use it to determine the compatibility of an input MMIF, accepting input that matches `major` level version. Thus this should solve the second problem above.

Okay.

> > - The `clams-python` package is set up to require a specific X.Y.Z version of the `mmif-python` package. I don't know what exactly the nature of this dependency is. Would it be feasible to use a range? At least for the patch-level and maybe also for the minor version.
>
> When the mmif-python and clams-python SDK are updated so that a CLAMS app can accept an input MMIF that's different in `minor` or `patch` level, the python dependency should not be relevant at all. And from the engineering perspective, I don't like the idea of having a range of versions as dependency. But we can re-consider this at any point. 

Not sure if I get this. Are you saying that the Python dependency between clams-python and mmif-python is only relevant at the major level? If clams-python version X requires mmif-python X.Y.Z it will use X.Y.Z and not X.A.B.

> Just to clarify the nature of version control between `mmif-python` and `clams-python`, `mmif-python` simply keeps the version of the spec as `mmif.__specver__` (as retrieved from the latest tag on the github at the build time)  and `clams-python` re-imports that as `clams.__specver__`. So with the change proposed at clamsproject/mmif-python#163, if a CLAMS app that is based on `clams-python==2.2.3` which in turn depends on `mmif.__version__==3.2.10 && mmif.__specver__==3.2.3` should be able to process any `3.[0-2].b` MMIF inputs. 

So `mmif.__version__` and `mmif.__specver__` are different things?

### Forward compatibility

> At this point, we have to think about forward compatibility. Namely, must this app be able to process a, for example, `3.7.x` MMIF input? Note that with proposal in this comment, once we reach 1.0.0, the only place the `minor` number gets a bump is *non-breaking* changes in `schema`. So the `schema==3.7.x` might have any number of optional fields added. But the `mmif-python==3.2.x` probably don't know how to handle those added fields, and possibly break with the higher version input. 

I think we do want that an app that uses 3.7.5 should be able to process input that is 3.7.6 or maybe even 3.8.0. And we want that for the sake of the sanity of all people putting apps into pipelines. Maybe we should think about separate functionality that builds on the spec versions and the `input` and `output` metadata to compute compatibility for a particular pipeline.

> The issue gets a bit more complicated when it comes to `0.x.y`, as within this range, `minor` and `patch` levels are conceptually mixed.

Yes, indeed. But that may give us an opportunity to experiment with what works and what doesn't when it comes to changes.

**Comment by keighrim** [comment]
> ### Version binding
> > > * The `clams-python` package is set up to require a specific X.Y.Z version of the `mmif-python` package. I don't know what exactly the nature of this dependency is. Would it be feasible to use a range? At least for the patch-level and maybe also for the minor version.
> > 
> > 
> > When the mmif-python and clams-python SDK are updated so that a CLAMS app can accept an input MMIF that's different in `minor` or `patch` level, the python dependency should not be relevant at all. And from the engineering perspective, I don't like the idea of having a range of versions as dependency. But we can re-consider this at any point.
> 
> Not sure if I get this. Are you saying that the Python dependency between clams-python and mmif-python is only relevant at the major level? If clams-python version X requires mmif-python X.Y.Z it will use X.Y.Z and not X.A.B.
> 

So a CLAMS app that uses any version of `clams-python` is able to accept any version of MMIF json as input. as `clams-python` does not check the version of a specification that an input MMIF is based on, at least for now (But MMIF files keep the spec version in its top metadata, so we can implement some kind of *rejection* in the `clams-python` in the future if we decide to).

Problem we and Kevin have now (clamsproject/mmif-python#163) is that, for example, when an app 
* is built upon `clams-python==0.4.0` that depends on `mmif-python==0.3.18` and `mmif.__specver__==0.3.2` in turn and
* is aimed to process some `TimeFrame` annotations and create a new view,

the following code snippet cannot retrieve any existing annotations from an input MMIF contains timeframes that have `@type` of `http://.../0.3.0/TimeFrame`. 
``` python
> mmif.__specver__ 
"0.3.2"
> mmif.vocabulary.AnnotationTypes.TimeFrame
"http://.../0.3.2/TimeFrame # this is a plain string hard-coded in an `enum` class"
> m = Mmif(input_mmif)
> m.find_all_views_contain(AnnotationTypes.TimeFrame)` 
[] # internal membership check just using string match, and thus fail to find any (because "*0.3.2*" != "*0.3.0*")
```
So in essence, this app does still run, it only fails to find any existing annotations and returns a new view with an empty list. 

#170 changes this by patching `mmif-python` so now that `find_all_views_contain(AnnotationTypes.TimeFrame)` call returns all views with `http://.../0.3.0/TimeFrame` annotations. Note that there's no change in `clams-python` code. What I meant in the previous comment is that once we decide rules of compatibility between spec versions and they are implemented in the `mmif-python==x.y.z` package (and we already have an implementation in #170), any future `clams-python` based apps should not suffer from clamsproject/mmif-python#163 issue. So in that sense, the dependency between `clams-python` and `mmif-python` is irrelevant to the rules of compatibility between spec versions at all. In other words, we do not impose any version binding between `clams-python` and `mmif-python` or `clams-ptyhon` and MMIF spec. 

> So `mmif.__version__` and `mmif.__specver__` are different things?

Yes, `mmif.__version__` keeps the SDK version and `mmif.__specver__` keeps the specification version the SDK based on. 
> 
> ### Forward compatibility
> I think we do want that an app that uses 3.7.5 should be able to process input that is 3.7.6 or maybe even 3.8.0. And we want that for the sake of the sanity of all people putting apps into pipelines. Maybe we should think about separate functionality that builds on the spec versions and the `input` and `output` metadata to compute compatibility for a particular pipeline.
> 

Now let's get back to the above example with following additional assumptions. 
* Spec 0.3.2 adds an optional property `frameType` to `TimeFrame` type. 
* The app actually wants `TimeFrame`s with ``frameType=="speech"`

(of course this difference must be a *breaking* change, but bear with me for the example)

Now suppose the app has code like this; 
``` python 
new_view = m.new_view()
self.sign_view(new_view)
for annotation in old_view.get_annotations(at_type=AnnotationTypes.TimeFrame):
  if "frameType" in annotation and annotation.frameType == "speech" 
    new_view.add_annotation(self.do_something(annotation, start_frame))
```

If we pass a MMIF with `http://.../0.3.0/TimeFrame`, all existing `TimeFrame` annotations fail to pass the *if* test and as a result new_view will have an empty annotation list. But that is exactly what we expect. 

Now with a non-breaking change 
* Spec 0.3.3 adds an optional property `averagePitch` to `TimeFrame` type. (completely fictional)
* The app isn't updated since spec 0.3.3 release

If we pass a MMIF with `http://.../0.3.3/TimeFrame`, the app still works as intended regardless of some `TimeFrame`s have `averagePitch` set or not. So forward compatibility with non-breaking changes seems to be okay. 

> Yes, indeed. But that may give us an opportunity to experiment with what works and what doesn't when it comes to changes.

Yup. Totally agree.

**Comment by keighrim** [comment]
I started a PR to update documentation on versioning with respect to our discussion so far.

**Comment by keighrim** [comment]
I also added [this table](https://clams.ai/mmif-python/latest/target-versions.html) and will add a similar table to clams-python documentation soon. I hope them help clarifying some confusion around the version compatibility.

**Comment by keighrim** [comment]
With the distinction between forward and backward compatibility in mind, I read through the semantic versioning documentation once more. Here's my latest interpretation of the sem-ver under our circumstance, and an updated proposal for future versioning of MMIF and `mmif-python`. 

---

So the sem-ver says; 

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> 1. MAJOR version when you make incompatible API changes,
> 1. MINOR version when you add functionality in a backwards compatible manner, and
> 1. PATCH version when you make backwards compatible bug fixes.

So considering our discussion on breaking-ness and forward/backward compatibility, my interpretation is ( `_` means any number)
1. MAJOR update breaks backward compatibility, meaning an app that expects `a._._` must not accept an input `b._._` if `a!=b`.
2. MINOR update is backward compatible, but breaks forward compatibility, meaning (given `a<b`),
    1. an app that expects `x.a._` must not accept an input `x.b._`. 
    2. an app that expects `x.b._` should be able to process an input `x.a._`. 
3. PATCH update does not break in either direction, meaning an app that expects `x.y.z` should be able to process an input `x.y._`.

Patches I wrote for clamsproject/mmif-python#170 and clamsproject/mmif-python#178 are only implementing the above PATCH-level compatibility. 

---

Now on versioning ;
### `schema` changes
This is very clear at a high level. *If a schema `x` does not validate an input `y` (`y` is invalid under `x`), version `x` is not compatible with version `y`.* But when we go into actual examples, it gets pretty complicated pretty quickly. So we might want to take on a case-by-case basis, as @marcverhagen pointed out before. Here are some examples which do NOT cover all scenarios; 
1. Backward-breaking (MAJOR): an input `y` is invalid under a schema `x` when `y` is older.
    1. Addition of a required field where `additionalProperties=false`: an older input `y` do not have that required field, so not a valid input.
    1. Renaming of a required field: This should be seen as an addition and a deletion. Regardless of the `additionalProperties` value, doing two actions must be backward breaking. 
    1. Modification of internal structure of a required field: an older input `y` have a different name or structure, so not a valid input. 
    1. Changing `additionalProperties` from `true` to `false`: an older input `y` might have arbitrary fields that are not allowed any more. 
2. Forward-breaking (MINOR): an input `y` is invalid under a schema `x` when `y` is newer.
    1. Deletion of a required field where `additionalProperties=true`: an newer input `y` might be missing that delete field that were required in the past. `y` is invalid. 
    1. Changing `additionalProperties` from `false` to `true`: a newer input `y` might contains arbitrary fields that were not allowed in the past. `y` is invalid.
3. Non-breaking (PATCH): schema `x` can validate input `y` even though `x!=y`.
    1. An addition of a field where `additionalProperties=true`: a newer input might have that new field, it was optional in the past anyway. Input is valid. 
    1. Changes in field metadata (title, description, etc.)

### `vocab` changes
This can be much simpler than `schema` in that there's only so many possible changes because we don't have `required-ness` in the type metadata/properties and all types are basically having `additionalProperties=true`. The criteria here would be that *if an app of `x` raises errors while processing input of `y`, version `x` and version `y` are incompatible.* Here's possible changes that can occur: 
1. Addition of a new type: 
    1. an older input doesn't know about this type, so a newer app just finds nothing. It doesn't seem the most intuitive to mark an app that's expecting type `X` as input compatible with all other apps that does not generate type `X` annotations at all. But this shouldn't throw any error. So not backward-breaking
    1. an older app doesn't know about this type, so this new type in a newer input is irrelevant to the app. Not forward-breaking.
1. Deletion of a type
    1. an older input might have this type, and a newer app thinks it is just a custom type. Not backward-breaking
    1. a newer input never have this type, so an older app finds nothing. Not farward-breaking.
1. Rename of a type
    1. this is a deletion and then an addition. Both are not breaking, so rename is also not breaking. 
1. Addition of a property to a type
    1. an older input doesn't have this property, so when a newer app tries to look for the property and this property is sort of "required", then it will raise `KeyError` or its equivalent. However, at the moment, we don't have a mechanics to mark a type property as required and we're treating all properties as optional except for the `id`. So this is not backward-breaking. 
    1. a newer input does have this property, but an older app never tries to look for this property. Not forward-breaking. 
1. Deletion of a property to a type
    1. an older input might have this property, a newer app never tries to look for the property. Not backward-breaking.
    1. a newer input does not have this property, but an older app tries to look for this property. This can raises `KeyError` maybe in the future when we have `required-ness` in the vocab, but now it is not breaking. 
1. Modification (including rename) of a property to a type
    1. an older input might have a different nested structure or name that a newer app looks for, raises `KeyError` (wrong name) or `ValueError` (wrong data type or structure). **Backward-breaking**.
    1. an older app looks for a different name or expects a different structure. **Forward-breaking**. 
1. Moving a type in the hierarchy
    1. Based on the new path to the root, this can result in a different property structure because inherited properties. Thus should be treated as modification of properties of a type. 

So using the breaking-ness above, we can say: 
1. Backward-breaking (MAJOR):
    1. property modification
    1. type movement with changing inheritance
1. Forward-breaking (MINOR):
    1. None
1. Non-breaking (PATCH):
    1. type addition 
    1. type deletion 
    1. type rename
    1. property addition
    1. property deletion 
    1. type movement without changing inheritance
    1. typo, description

---

For now since we don't use `major` numbers at all, I think we should use `minor` bump for both backward and forward breaking changes. So with that said, current implementation of version checking (`patch` only) is actually enough until we start using `major` versions. Let me know how you think.

**Comment by marcverhagen** [comment]
Thanks for coming up with that hierarchy.

I will look into this later, once I have worked more on the NER application. And I think we should table this till we have done more work on applications.

One comment is that I have some trouble with the terms forward-breaking and backward-breaking.

**Comment by keighrim** [comment]
Some random notes from today's meeting with Marc

1. Having an umbrella MMIF version for all vocab types is the root cause of issues like https://github.com/clamsproject/mmif-python/issues/163 and confusions like https://github.com/clamsproject/mmif-python/issues/200. 
2. Instead, having separated versioning schemes for all individual types will allow developers to use a free combination of any versions of any annotation types in any order, not "bound" to the MMIF spec-ver that the SDK is based/targeting on.
3. Namely, individuated versioning makes tying a "snapshot" of the hierarchy into an umbrella MMIF version mostly, if not completely, meaningless. And thus, with that model, vocab changes no longer change MMIF version numbers, they should be totally independent. (emancipation of the "CLAMS" vocab from the MMIF)
4. Under individuated versioning of CLAMS types, the HTML representation of a type hierarchy actually can (and probably must) be built dynamically with a "selected" set of versions and types. The type/version selection is done by a user (most likely an app developer) and can lead to a possible conflict in parent-child relations (this is not very likely though). 
5. Our MMIF jsonscheme is already pretty mature and stable, and so is the associated specification documentation.

**Comment by keighrim** [comment]
Self-answering my own questions based on the in-person discussion today. 

> * Which version of `TypeA` should the above statement default to, or should there be a default at all?

It should default to the version shipped with the SDK, just like what's currently done. 

> * How can a developer research for a version or a range of versions that they want to use as input to their app, or how much of it is the developer's responsibility?

The developer should not have to research a version or a range of versions to support during developing an app. If there's ever a need to handle an old version of vocab type in a new version, there must be a separate "converter" app to be developed and pipelined together. 

> * How much, if any, should the SDK support this type-version continuum matching (in `annotate()`), or should the type coercion be totally a burden for the developer (in `_annotate()`)?

None, based on the above answer.

> * If it's totally the developers' burden, does type versioning mean anything at all in terms of type-checking? Can developers just check the structure of the input annotation object and process only when there's enough (and properly structured) information? (type versioning is still meaningful in terms of bookkeeping, of course)

No longer relevant, as we won't support at_type version specification in `appmetadata`.

**Comment by keighrim** [comment]
#197 implements an individuated versioning of annotation types. However, regarding these points suggested in the above, 
> Namely, individuated versioning makes tying a "snapshot" of the hierarchy into an umbrella MMIF version mostly, if not completely, meaningless. And thus, with that model, vocab changes no longer change MMIF version numbers, they should be totally independent. (emancipation of the "CLAMS" vocab from the MMIF)

> Under individuated versioning of CLAMS types, the HTML representation of a type hierarchy actually can (and probably must) be built dynamically with a "selected" set of versions and types. 

The PR implements *emancipation* of the versions of individual annotation types from the MMIF version, NOT the *emancipation* of the "CLAMS" vocab from the MMIF. Namely, changes in a vocab annotation type still should bump the MMIF version. In other words, it does NOT aim at "dynamic construction" of vocab tree. Instead, there are sill "static" hierarchies bound to each MMIF version, where is a tree of which nodes are annotations types and edges are relations between those annotation types. The withdrawal of the idea of the dynamic tree construction is based on our earlier decision to give up supporting the "free-range" I/O specification in the SDK (https://github.com/clamsproject/mmif-python/issues/163#issuecomment-1463244039)

---

A few things we still need to decide; 

1. the format of at_type version appearing in the URI: #194 changed the order of components in an annotation type URI. At that time, there was no individuated versioning, so the raw MMIF version is used to construct the URI. The versioning scheme proposed in this PR is just using a single integer. Here's some screenshot on how they are rendered as a URI and on the website. 
(ignore `CLAMS Vocabulary x.y.z` in the header banner, we'll come back to that number in the next bullet point)
![image](https://user-images.githubusercontent.com/9062727/229928857-c6750021-48a0-4de7-b1ad-025e6975ca24.png)
![image](https://user-images.githubusercontent.com/9062727/229928979-2524f41e-9052-49f4-9b30-2842b18486d0.png)
And I don't like how it's rendered. Most of all, I think the single integer format, at a glance, doesn't look like a version number. Alternatively, we can append a simple prefix like 
![image](https://user-images.githubusercontent.com/9062727/229928716-62a102fb-d84b-43c2-8848-5bb90e342b54.png)
![image](https://user-images.githubusercontent.com/9062727/229930753-f6990ee8-b5f8-46bf-8b04-32dac32446e9.png)
, which I like better. But we have to be careful to make the re-formatting process transparent (and preferably re-usable) so that downstream applications, including `mmif-python`, can follow the exact process.
Or we can use a completely different version format. For instance, I [once suggested](https://github.com/clamsproject/mmif-python/issues/163#issuecomment-1433995719) re-using the MMIF version number. 
1. A problem described in the following comment (regarding `CLAMS Vocabulary x.y.z` number in the header banner)
https://github.com/clamsproject/mmif/blob/e021331c4710324f19475ca24206a4fd4f471f1e/build.py#L181-L198
1. Which level in the MMIF semver should a change in a vocab annotation type should increase? In a previous comment (https://github.com/clamsproject/mmif/issues/14#issuecomment-860076628), I suggested the following: 
    > So using the breaking-ness above, we can say: 
    > 1. Backward-breaking (MAJOR):
    >     1. property modification
    >     1. type movement with changing inheritance
    > 1. Forward-breaking (MINOR):
    >     1. None
    > 1. Non-breaking (PATCH):
    >     1. type addition 
    >     1. type deletion 
    >     1. type rename
    >     1. property addition
    >     1. property deletion 
    >     1. type movement without changing inheritance
    >     1. typo, description

    However, since the "incompatibility" of vocab types will not break anything once we implement failing of type checking as *warnings* instead of errors. I'd like to suggest any changes in vocab items (which result in a new hierarchy tree due to replacement of one element in the tree - e.g. replacing `TimeFrame/v1` with `TimeFrame/v2`) will only bump the patch level number in the main MMIF version.

**Comment by keighrim** [comment]
On top of those, we need to map from current vocab items to `v1`. Since all but one vocab types (the capital `Annotation` type #191) have [not been updated for ~2 years](https://github.com/clamsproject/mmif/commits/develop/vocabulary/clams.vocabulary.yaml), or since 0.4.0, and all of currently available apps are based on [MMIF >0.4.0](https://github.com/clamsproject/clams-apps/blob/7a8b6d4b4d6edf01ad5bcbe41288c41bebb8f3cf/README.md), here's what I'm going to do 
* all `http://mmif.clams.ai/vocab/0.[1-3].*/*` types are obsolete and won't be mapped to anything. 
* all `http://mmif.clams.ai/vocab/0.4.*/*` types will be mapped to new type `v1`, except for
* `http://mmif.clams.ai/vocab/0.4.2/Annotation` types will be mapped to `http://mmif.clams.ai/vocab/Annotation/v2`
* Next MMIF release (`0.4.3`) will have a vocabulary consists of `Annotation/v2` and all other `AtType/v1`s.

**Comment by keighrim** [comment]
The herder/head components of individual type pages are re-done in a new commit (b50d8db01cd1ebfc8730db11d274c83f7e29e821 in #198). 
Here's a screenshot of how it looks. 


![Screenshot 2023-04-14 at 10 16 45](https://user-images.githubusercontent.com/9062727/232070702-7018405b-45d0-4e8a-a61a-5330982f1bbb.png)

**Comment by keighrim** [comment]
As annotation types versioning is newly introduced, most of comments prior to 2023 are now outdated and incorrect. I added a cleaned-up version of our latest versioning strategies and what "breaking" means in terms of CLAMS pipelines under the updated versioning as [a part of MMIF documentation in the PR](https://github.com/clamsproject/mmif/pull/197/files#diff-c1cc304e56dd03cbb5cf65c32d83c9b7ab069800b2ce780249ac87d320e76789). 

spoiler alert: MMIF versions are now **irrelevant** to type checking in pipelines.

**Comment by keighrim** [comment]
Finally, closed via #197.
